imageLib.ml

```diff
@@ -33,13 +33,14 @@ module JPG = ImageJPG
 module GIF = ImageGIF
 
 let convert fn fn' =
-  let cmd = Printf.sprintf "convert %s %s" fn fn' in
-  let ret = Sys.command cmd in
-  if ret <> 0 then assert false
+  let ret =
+    Unix.create_process "convert" [| "convert"; "--" ; fn ; fn'|]
```

Comment:
Why the `"--"` here?

Reply:
Because otherwise filenames starting with `-` and `--` will be interpreted as command-line arguments. `--` is the GNU way to say "ok, now I'm done with giving you switches and flags, the rest are literal values".
Example:
```shell
# make a file called '-F':
$ echo hi | tee -- -F
# list the number of lines in -F
$ wc -l -- -F
1 -F

# without -- this gives:
$ echo hi | tee -F
tee: invalid option -- 'F'
Try 'tee --help' for more information.
$ wc -l -F
wc: invalid option -- 'F'
Try 'wc --help' for more information.
```

Reply:
I know this is fairly standard, but (at least on my machine) the `convert` command does not support the `--` option.

Reply:
Oh, ugh, I guess that to support OSX / BSD we should really just verify that the path doesn't start with a `-`. I'll try to update the code.

Reply:
`convert` seems to ignore the first character completely:
```ocaml
Unix.create_process "convert" [|"convert";"-version"|] Unix.stdin Unix.stdout Unix.stderr;;
Unix.create_process "convert" [|"convert";" -version"|] Unix.stdin Unix.stdout Unix.stderr;;
Unix.create_process "convert" [|"convert";"\n-version"|] Unix.stdin Unix.stdout Unix.stderr;;
Unix.create_process "convert" [|"convert";"A-version"|] Unix.stdin Unix.stdout Unix.stderr;;
Unix.create_process "convert" [|"convert";"B-version"|] Unix.stdin Unix.stdout Unix.stderr;;
```

@hannesm Do you know if there's a sane way to do this on BSD?

Reply:
the `--` is the usual `end of options` marker (read [more](https://unix.stackexchange.com/questions/147143/when-and-how-was-the-double-dash-introduced-as-an-end-of-options-delimiter)) -- on my FreeBSD and tcsh system, I get:

```bash
# convert -version
Version: ImageMagick 6.9.9-28 Q16 amd64 2018-06-24 http://www.imagemagick.org
Copyright: © 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Features: Cipher DPC Modules 
Delegates (built-in): bzlib cairo fftw fontconfig freetype gvc jbig jng jp2 jpeg lcms lqr ltdl lzma png raw rsvg tiff webp wmf x xml zlib

# convert -- -version
hannes: unable to open image `-version': No such file or directory @ error/blob.c/OpenBlob/2720.
hannes: no decode delegate for this image format `' @ error/constitute.c/ReadImage/504.
hannes: missing an image filename `-version' @ error/convert.c/ConvertImageCommand/3255.
```

so, it is a feature of `getopt`, and not documented in all man pages.  I've been using `--` since several years on various UNIX systems, and so far didn't get into any trouble. :)

Reply:
Hmm, it works on my `convert` (on debian).
This seems to suggest it does handle `--` somehow? https://github.com/ImageMagick/ImageMagick/blob/master/MagickWand/convert.c#L636

Will try to investigate some more...

Reply:
Yes, that's what I thought looking at their code. What version do you run? Mine is `7.0.7-38`. @hannesm seems to have a slightly older version, maybe this is a `convert` bug introduced in between `6.9.9-28` and `7.0.7-38`.

Reply:
```
Version: ImageMagick 6.9.9-34 Q16 x86_64 20180122 http://www.imagemagick.org
```

Reply:
OK, I reported this [here](https://github.com/ImageMagick/ImageMagick/issues/1187). Feel free to add on that. For now, I suggest we remove the `--`. We do not use filenames starting with a dash that often, do we?

Reply:
@rlepigre without the `--`, I think you are also unable to convert a file like `a-version` or `b-version`? Anyway, I'll remove the `--` so we can get this merged, but I still think that calling out to `convert` is dangerous.

Reply:
@cfcs of course you are right, but what can we do about it right now? Maybe you should add something about `a-version` and related stuff to the issue I opened, I completely forgot about this part. This cannot be the expected behaviour. It seems that they check the second character for handling options with double dash, but in that case don't look at the first character.